linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -4212,27 +4222,65 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
     } else if (mutable) {
       withDedicatedVar(RefinedType(declaredType))
     } else {
-      val refinedType = value.tpe
+      def computeRefinedType(): RefinedType = bindingName match {
+        case _ if value.tpe.isExact || declaredType == AnyType =>
```

Comment:
To check my understanding: The reason we keep the `declaredType` around (in `Binding`) is to be able to properly do the coercions of long values?

Reply:
There are a few reasons to use the `declaredType`:
* The Long coercions, indeed
* When the binding is mutable, we must always use the declared type, and not a refinement of the value that happens to be used to initialize it
* And now specifically for this refinement, when the declared type is in fact more precise than `value.tpe`.